OrbitVulkanLayerClientGgp/DispatchTable.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitVulkanLayerClientGgp/DispatchTable.h"
+
+#include <string.h>
+
+#include "OrbitBase/Logging.h"
+#include "absl/base/casts.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "vulkan/vk_layer.h"
+#include "vulkan/vk_layer_dispatch_table.h"
+#include "vulkan/vulkan.h"
+
+namespace {
+// use the loader's dispatch table pointer as a key for dispatch map lookups
+template <typename DispatchableType>
+void* GetKey(DispatchableType inst) {
+  return *absl::bit_cast<void**>(inst);
+}
+}  // namespace
+
+void DispatchTable::CreateInstanceDispatchTable(const VkInstance& instance,
+                                                const PFN_vkGetInstanceProcAddr& gpa) {
+  // fetch our own dispatch table for the functions we need, into the next layer
+  VkLayerInstanceDispatchTable dispatch_table;
+  dispatch_table.GetInstanceProcAddr =
+      absl::bit_cast<PFN_vkGetInstanceProcAddr>(gpa(instance, "vkGetInstanceProcAddr"));
+  dispatch_table.DestroyInstance =
+      absl::bit_cast<PFN_vkDestroyInstance>(gpa(instance, "vkDestroyInstance"));
+  dispatch_table.EnumerateDeviceExtensionProperties =
+      absl::bit_cast<PFN_vkEnumerateDeviceExtensionProperties>(
+          gpa(instance, "vkEnumerateDeviceExtensionProperties"));
+
+  // store the table by key
+  instance_dispatch_[GetKey(instance)] = dispatch_table;
```

Comment:
The mutex are locked before the call. All the mutex logic is in the main file and transparent to this as it is handled from there. All calls are locked

Reply:
I am not 100% sure about the scope of your layer. If it stays quite simple, this could be fine. Otherwise, I'd suggest to move it to this class (see the other comment).

Reply:
This is being discussed in the other comment so I resolve this one.